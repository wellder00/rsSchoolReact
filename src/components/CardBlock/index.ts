export { default as CardBlock } from './CardBlock';
