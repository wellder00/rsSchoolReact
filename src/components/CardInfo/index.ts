export { default as CardInfo } from './CardInfo';
