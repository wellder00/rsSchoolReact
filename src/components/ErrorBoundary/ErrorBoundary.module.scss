.wrapper {
  box-sizing: border-box;
  background: linear-gradient(0deg, #040100, #d67a09, #bf9056);
  padding: 40px 15px;
  height: 100vh;
  font-family: 'Nosifer', sans-serif;
  text-align: center;
  @media (max-width: 700px) {
    font-size: 12px;
  }
  @media (max-width: 480px) {
    font-size: 10px;
  }
}

.title {
  color: rgb(141, 10, 10);
}

.wrapperPoem {
  margin: 40px 0;
}

.button {
  animation: pulse 2s infinite;
  cursor: pointer;
  box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
  background-color: #4caf50;
  padding: 10px;
  width: 300px;
  color: #fff;
  &:hover {
    animation-play-state: paused;
  }
}

@keyframes pulse {
  0% {
    transform: scale(1);
  }
  50% {
    transform: scale(1.1);
  }
  100% {
    transform: scale(1);
  }
}
