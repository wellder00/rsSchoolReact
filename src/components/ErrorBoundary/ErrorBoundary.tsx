import React, { Component, ReactNode } from 'react';

import styles from './ErrorBoundary.module.scss';

interface ErrorBoundaryProps {
  children: ReactNode;
}

interface ErrorBoundaryState {
  hasError: boolean;
  error: Error | null;
}

class ErrorBoundary extends Component<ErrorBoundaryProps, ErrorBoundaryState> {
  constructor(props: ErrorBoundaryProps) {
    super(props);
    this.state = { hasError: false, error: null };
  }

  static getDerivedStateFromError(error: Error) {
    return { hasError: true, error };
  }

  componentDidCatch(error: Error, errorInfo: React.ErrorInfo) {
    console.error(error, errorInfo);
  }

  reloadPage = () => {
    window.location.reload();
  };

  render() {
    if (this.state.hasError) {
      return (
        <div className={styles.wrapper}>
          <h3 className={styles.title}>Error Message: {this.state.error?.message}</h3>
          <div className={styles.wrapperPoem}>
            <p>In the realm of code, where errors roam,</p>
            <p>A guardian stands, it&apos;s ErrorBoundary&apos;s home.</p>
            <p>With watchful eyes, it scans the land,</p>
            <p>To catch the bugs, with a helping hand.</p>
          </div>

          <div className={styles.wrapperPoem}>
            <p>When errors arise, like a shadow they loom,</p>
            <p>ErrorBoundary steps in, dispelling the gloom.</p>
            <p>With grace and poise, it takes the lead,</p>
            <p>Catching those bugs, it&apos;s just what we need.</p>
          </div>
          <div className={styles.wrapperPoem}>
            <p>In verses of JavaScript, it scripts the play,</p>
            <p>Ensuring our apps run smoothly each day.</p>
            <p>So fear not the bugs, let ErrorBoundary be,</p>
            <p>The hero of code, setting errors free.</p>
          </div>
          <div className={styles.wrapperPoem}>
            <p>In React&apos;s embrace, it finds its place,</p>
            <p>A protector of code, with style and grace.</p>
            <p>When errors strike, it&apos;s ErrorBoundary&apos;s song,</p>
            <p>To keep our apps robust and strong.</p>
          </div>

          <button className={styles.button} onClick={this.reloadPage}>
            Reload Page
          </button>
        </div>
      );
    }
    return this.props.children;
  }
}

export default ErrorBoundary;
