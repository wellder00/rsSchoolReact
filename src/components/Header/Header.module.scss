@import '../../styles/variables.scss';

.wrapper {
  display: flex;
  justify-content: space-around;
  align-items: center;
  gap: 20px;
  padding: 20px 0;
  background: $card-color;
  width: 100%;
  @media (max-width: 1015px) {
    flex-direction: column;
  }
}

.searchWrap {
  display: flex;
  justify-content: center;
  gap: 20px;
  @media (max-width: 586px) {
    flex-direction: column;
  }
}

.titleImg {
  width: 250px;
  filter: drop-shadow(0 0 20px rgb(0, 5, 9));
}

.logo {
  width: 200px;
  filter: drop-shadow(0 0 20px red);
  @media (max-width: 1015px) {
    display: none;
  }
}
