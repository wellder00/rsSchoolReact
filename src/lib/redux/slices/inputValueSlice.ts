import { createSlice, PayloadAction } from '@reduxjs/toolkit';

type InputValueState = {
  value: string;
};

const initialState: InputValueState = {
  value: '',
};

const pokemonSlice = createSlice({
  name: 'inputValue',
  initialState,
  reducers: {
    saveItemValue(state, action: PayloadAction<string>) {
      state.value = action.payload;
    },
  },
});

export const { saveItemValue } = pokemonSlice.actions;

export default pokemonSlice.reducer;
