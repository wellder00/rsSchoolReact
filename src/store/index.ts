import { configureStore } from '@reduxjs/toolkit';
import inputValueReducer from './inputValueSlice';
import itemsAmountReducer from './itemsPerPageSlice';
import { pokemonApi } from './redux/pokemonApi';

const store = configureStore({
  reducer: {
    inputValue: inputValueReducer,
    itemsAmount: itemsAmountReducer,
    [pokemonApi.reducerPath]: pokemonApi.reducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(pokemonApi.middleware),
});

export default store;

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
