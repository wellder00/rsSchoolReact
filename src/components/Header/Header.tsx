import React, { useState, useEffect, useContext } from 'react';

import styles from './Header.module.scss';

import { InputSearch } from '../InputSearch';
import { Button } from '../Button';

import title from '../../assets/images/title.png';
import logo from '../../assets/images/logo.png';
import pokemonDataContext from '../../state/ContextPokemonData';
import { saveItemValue } from '../../store/inputValueSlice';
import { useAppDispatch, useAppSelector } from '../../Hooks/reduxHooks';

type Props = {
  findCharacter: (selectedCategory: string) => void;
  onSelectChange: (event: React.ChangeEvent<HTMLSelectElement>) => void;
};

const Header: React.FC<Props> = ({ findCharacter, onSelectChange }) => {
  const dispatch = useAppDispatch();
  const itemsAmount = useAppSelector((state) => state.itemsAmount.items);
  const inputValue = useAppSelector((state) => state.inputValue.value);
  const [hasError, setHasError] = useState(false);
  const PokemonDate = useContext(pokemonDataContext);

  //!!! Проверить эту функцию
  useEffect(() => {
    const storedData = localStorage.getItem('character');
    if (storedData) {
      const parsedData = JSON.parse(storedData);
      dispatch(saveItemValue(parsedData));
    }
  }, []);

  const handleInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    dispatch(saveItemValue(event.target.value));
  };

  const handleSearch = () => {
    findCharacter(inputValue);
    dispatch(saveItemValue(''));
  };

  const handleMakeError = () => {
    setHasError(true);
  };
  if (hasError) {
    throw new Error('Everything is broken, everything is destroyed');
  }

  return (
    <div className={styles.wrapper}>
      <img className={styles.titleImg} src={title} alt="title" />
      <div className={styles.searchWrap}>
        {PokemonDate?.count && (
          <select onChange={onSelectChange} className={styles.select} value={itemsAmount}>
            <option value="10">10</option>
            <option value="20">20</option>
            <option value="30">30</option>
          </select>
        )}

        <InputSearch
          className={'search'}
          placeholder={'search'}
          onChange={handleInputChange}
          getInputValue={handleSearch}
        />

        <Button className={'search'} onClickFunction={handleSearch}>
          SEARCH
        </Button>

        <Button className={'error'} onClickFunction={handleMakeError}>
          TRY MAKE ERROR
        </Button>
      </div>
      <img className={styles.logo} src={logo} alt="logo" />
    </div>
  );
};

export default Header;
