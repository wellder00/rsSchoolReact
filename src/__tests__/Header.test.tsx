import { render, screen, fireEvent } from '@testing-library/react';
import { Header } from '../components/Header';
import { ChangeEvent } from 'react';
import * as reduxHooks from 'react-redux';
import * as actions from '../store/inputValueSlice';
import '@testing-library/jest-dom';
import { ErrorBoundary } from '../components/ErrorBoundary';

jest.mock('react-redux');

const mockedSelector = jest.spyOn(reduxHooks, 'useSelector');
const mockedDispatch = jest.spyOn(reduxHooks, 'useDispatch');
describe('Header', () => {
  test('renders Header component', () => {
    mockedDispatch.mockReturnValue(jest.fn());
    mockedSelector.mockReturnValue('');
    const component = render(
      <Header
        onSelectChange={(event: ChangeEvent<HTMLSelectElement>): void => {
          throw new Error('Function not implemented.');
        }}
      />
    );
    expect(component).toMatchSnapshot();
  });
  test('find text', () => {
    mockedSelector.mockReturnValue('some text');
    const component = render(
      <Header
        onSelectChange={(event: ChangeEvent<HTMLSelectElement>): void => {
          throw new Error('Function not implemented.');
        }}
      />
    );
    expect(screen.getByPlaceholderText('search')).toBeInTheDocument();
  });
  test('show dishatch actions', () => {
    const dispatch = jest.fn();
    mockedDispatch.mockReturnValue(dispatch);

    const mockedSaveValue = jest.spyOn(actions, 'saveItemValue');

    render(
      <Header
        onSelectChange={(event: ChangeEvent<HTMLSelectElement>): void => {
          throw new Error('Function not implemented.');
        }}
      />
    );
    const searchButton = screen.getByText('SEARCH');
    fireEvent.click(searchButton);
    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(mockedSaveValue).toHaveBeenCalled();
  });
  test('get Error actions', () => {
    const consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation(() => {});
    render(
      <ErrorBoundary>
        <Header
          onSelectChange={(event: ChangeEvent<HTMLSelectElement>): void => {
            throw new Error('Everything is broken, everything is destroyed');
          }}
        />
      </ErrorBoundary>
    );
    const errorButton = screen.getByText('TRY MAKE ERROR');
    fireEvent.click(errorButton);
    expect(screen.getByText('Reload Page')).toBeInTheDocument();
    consoleErrorSpy.mockRestore();
  });
});
