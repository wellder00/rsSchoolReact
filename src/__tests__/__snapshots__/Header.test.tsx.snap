// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders Header component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="wrapper"
      >
        <img
          alt="title"
          class="titleImg"
          src="title.png"
        />
        <div
          class="searchWrap"
        >
          <select
            class="select"
          >
            <option
              value="10"
            >
              10
            </option>
            <option
              value="20"
            >
              20
            </option>
            <option
              value="30"
            >
              30
            </option>
          </select>
          <input
            class="search"
            data-testid="input-test-id"
            placeholder="search"
            type="text"
            value=""
          />
          <button
            class="search button"
          >
            SEARCH
          </button>
          <button
            class="error button"
          >
            TRY MAKE ERROR
          </button>
        </div>
        <img
          alt="logo"
          class="logo"
          src="logo.png"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="wrapper"
    >
      <img
        alt="title"
        class="titleImg"
        src="title.png"
      />
      <div
        class="searchWrap"
      >
        <select
          class="select"
        >
          <option
            value="10"
          >
            10
          </option>
          <option
            value="20"
          >
            20
          </option>
          <option
            value="30"
          >
            30
          </option>
        </select>
        <input
          class="search"
          data-testid="input-test-id"
          placeholder="search"
          type="text"
          value=""
        />
        <button
          class="search button"
        >
          SEARCH
        </button>
        <button
          class="error button"
        >
          TRY MAKE ERROR
        </button>
      </div>
      <img
        alt="logo"
        class="logo"
        src="logo.png"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
